packages/ai-chat/src/browser/session-summary-variable-contribution.ts

```diff
@@ -0,0 +1,115 @@
+        const selection = await this.quickInputService.showQuickPick(items);
+        return selection?.id;
+    }
+
+    protected async provideCompletionItems(
+        model: monaco.editor.ITextModel,
+        position: monaco.Position,
+        matchString?: string
+    ): Promise<monaco.languages.CompletionItem[] | undefined> {
+        const context = AIVariableCompletionContext.get(SESSION_SUMMARY_VARIABLE.name, model, position, matchString);
+        if (!context) { return undefined; }
+        const { userInput, range, prefix } = context;
+        return this.getItems().filter(candidate => QuickPickItem.is(candidate) && candidate.label.startsWith(userInput)).map(({ label, id }: QuickPickItem) => ({
+            label,
+            kind: monaco.languages.CompletionItemKind.Class,
+            range,
+            insertText: `${prefix}${id}`,
+            detail: id,
+            filterText: userInput,
+        }));
+    }
+
+    protected getItems(): QuickPickItemOrSeparator[] {
+        return [
+            ...(this.summaries.size ? [{ type: 'separator', label: 'Recent Summaries' }] satisfies QuickPickSeparator[] : []),
+            ...Array.from(this.summaries.entries(), ([id, { label }]) => ({
+                type: 'item',
+                label,
+                id
+            })) satisfies QuickPickItem[],
+            ...(this.summaries.size ? [{ type: 'separator', label: 'Other Sessions' }] satisfies QuickPickSeparator[] : []),
+            ...this.chatService.getSessions()
+                .filter(candidate => !this.summaries.has(candidate.id) && candidate.model.getRequests().length)
+                .map<QuickPickItem>(session => ({ type: 'item', label: session.title || session.id, id: session.id }))
+        ];
+    }
+
+    canResolve(request: AIVariableResolutionRequest, context: AIVariableContext): MaybePromise<number> {
+        return request.variable.id === SESSION_SUMMARY_VARIABLE.id ? 10000 : -5;
+    }
+
+    async resolve(request: AIVariableResolutionRequest, context: AIVariableContext): Promise<ResolvedAIContextVariable | undefined> {
+        if (request.variable.id !== SESSION_SUMMARY_VARIABLE.id || !request.arg) { return; }
+        const existingSession = this.chatService.getSession(request.arg);
+        const existingSummary = this.summaries.get(request.arg);
+        const newSummaryPossibleAndNecessary = !!existingSession && (!existingSummary || existingSummary.length !== existingSession.model.getRequests().length);
+        try {
+            const value = newSummaryPossibleAndNecessary ? await this.summarizeSession(existingSession) : existingSummary?.summary;
+            return value ? { ...request, value, contextValue: value } : undefined;
+        } catch (err) {
+            console.warn('Error retrieving chat session summary for session', request.arg, err);
+            return;
+        }
+    }
+
+    protected async summarizeSession(session: ChatSession): Promise<string> {
```

Comment:
Question 4: We discussed to create the summaries not on the fly, but at that moment, when the summary is created. For the current use case, this would mean if the user creates a summary from a chat and continues the summarized chat afterwards, the summary would not be automatically updated, but I would not expect this. fine?

Reply:
4. fine

Reply:
No updates are expected, though if the summary is stored as a file, the user could go and modify it.

Reply:
@colin-grant-work I think the summaries are still created only when the variable is resolved. I would create them when the user clicks the button. Is there any reason to only create them lazily? The lazy creation also introduces some weirdness: If I summarize a chat and I have a folder for persistence is configured, nothing happens until I either enter something in the new chat or click on the summary variable to see its content. This weird behavior would be gone if we summarize directly on clicking the action.

Reply:
Yes, also added as a todo.

Reply:
I see, that was this "(or create it earlier)" :-)